### ISSUE TITLE
Replace annonymous class with lambda expression

### ISSUE THREAD

**Issue by exoego** [circe/circe#1168]
There are lots (100+) of annonymous class instantiation of SAM (Single Abstract Method) types in circe code base.
https://github.com/circe/circe/blob/23af4778f1bf5b8ba8cfc93a6c32b83350a1dffc/modules/shapes/src/main/scala/io/circe/shapes/CoproductInstances.scala#L7-L9

If I understand correctly, replacing them with lambda expression can eliminate `.class` files for anonymous classes in Scala 2.12, since Scala 2.12 can use bytecode `invokedynamic` for lambdas.
Therefore refactoring with lamda expression will result in 

* Faster compilation
* Smaller JAR file
* Faster runtime performance

Example of refactoring:

```scala
// Before
implicit final val decodeCNil: Decoder[CNil] = new Decoder[CNil] {
  def apply(c: HCursor): Decoder.Result[CNil] = Left(DecodingFailure("CNil", c.history))
}

// After
implicit final val decodeCNil: Decoder[CNil] =
  (c: HCursor) => Left(DecodingFailure("CNil", c.history))
```

**Comment by travisbrown** [comment]
Unfortunately the compiler won't actually use `LambdaMetafactory` for instances of `Decoder` defined with the SAM syntax, so even if we were willing to use the experimental SAM support for 2.11 in circe (which I'm not sure about—I seem to remember running into problems with it), it wouldn't really help us out in this respect, since we'd still get the anonymous class representation.

At some point a couple of years ago I spent some time working through what the 2.12 compiler does in cases like this, but it seemed like a confusing mess and to be honest I don't really remember why `Decoder` doesn't get the LMF treatment. Even this simplified version doesn't

```scala
trait Decoder[A] {
  def apply(input: String): Option[A]
  final def orElse(input: String, a: A): A = apply(input).getOrElse(a)
}

object Test {
  val stringDecoder: Decoder[String] = (input: String) => Some(input)
}
```
This will still end up with an anonymous class for the `stringDecoder` implementation. I don't know why—from my reading of the (scant) documentation it seems like this `Decoder` is pretty clearly a "simple interface" that should qualify. Remove the `orElse` and you'll get the LMF version, but that's not really an option for us.

I've also tried benchmarking the SAM syntax vs. `new Decoder` vs. `Decoder.instance` before, and `new Decoder` has been far enough ahead that I thought we should standardize on it in the circe implementation (although I'd expect most users to choose either `Decoder.instance` or the SAM syntax). That might change someday, and I should probably dig up that benchmark and run it on 2.13.0, but at least as far as LMF vs. anonymous class goes, it seems like 2.13.0 is the same as 2.12.8.
